src/daft-micropartition/src/micropartition.rs

```diff
@@ -504,6 +506,40 @@ impl MicroPartition {
             unreachable!()
         }
     }
+
+    pub(crate) fn add_monotonically_increasing_id(
+        &self,
+        partition_num: u64,
+        column_name: &str,
+    ) -> DaftResult<Self> {
+        let io_stats = IOStatsContext::new("MicroPartition::add_monotonically_increasing_id");
+        let tables = self.tables_or_read(io_stats)?;
+
+        let tables_with_id = tables
+            .iter()
+            .scan(0u64, |offset, table| {
+                let table_with_id =
+                    table.add_monotonically_increasing_id(partition_num, *offset, column_name);
+                *offset += table.len() as u64;
+                Some(table_with_id)
+            })
+            .collect::<DaftResult<Vec<_>>>()?;
+
+        let mut schema_with_id_index_map = self.schema.fields.clone();
+        schema_with_id_index_map.insert(
+            column_name.to_string(),
+            Field::new(column_name, DataType::UInt64),
+        );
+        let schema_with_id = Schema {
+            fields: schema_with_id_index_map,
+        };
+
+        Ok(Self::new_loaded(
+            Arc::new(schema_with_id),
+            Arc::new(tables_with_id),
+            self.statistics.clone(),
```

Comment:
Should we calculate new statistics for the new column and add it in here? 

Maybe this is more of a @samster25 question: if we create new columns should we also be calculating statistics and amending the stats in our MicroPartition?

Reply:
I think it should be fine! If theres a missing columns, it returns Missing by Default. Also this already materializes the source files, so there wouldn't be much of a benefit.